atc/exec/set_pipeline_step.go

```diff
@@ -19,22 +19,26 @@ import (
 	"github.com/concourse/concourse/atc/db"
 	"github.com/concourse/concourse/atc/exec/artifact"
 	"github.com/concourse/concourse/atc/exec/build"
+	"github.com/concourse/concourse/atc/policy"
 	"github.com/concourse/concourse/atc/worker"
 	"github.com/concourse/concourse/tracing"
 	"github.com/concourse/concourse/vars"
 )
 
+const ActionRunSetPipeline = "RunSetPipeline"
```

Comment:
Nice!

@evanchaoli @cludden Follow-up: do you think it would be valuable to have this single action run in both paths (set via API vs set via step)? Then the policy writer would only need to implement one policy. Right now it seems easy to forget to do one or the other.

If so, I think this could be renamed to just `SetPipeline`. Assuming the current name is from "run `set_pipeline` step".

A deeper follow-up topic: I wonder if it would be safer to move away from generic API policy actions and instead implement specific actions like this. It seems like API-level policy enforcement may not be the most flexible thing going forward, since there are ways for an action to occur without going through the API. This is a trend that I expect to continue as we make Concourse more capable of automating itself.

Are there other common situations where we want to vet the API request, including the request payload? Can those be turned into more specific actions as well?

I recognize that being able to enforce policies for all API endpoints is super flexible and a cheap way to make the policy enforcement really powerful, but it works by casting a pretty large net, and it may become really fiddly over time. 🤔 I'm just asking these questions now while it's still experimental.

Thanks!

Reply:
FWIW, I agree. My original PR intended to keep the same behavior regardless of where `SaveConfig` is invoked (cli, api, set_pipeline step). A quick fix to move this PR forward would be to rename `SaveConfig` to `SetPipeline` and remove the associated HTTP method from the check input in both places (similar to `UseImage`). A medium effort fix, most likely outside the scope of this PR, would be to remove the HTTP method filters entirely, and simply rely on action based filters. A larger refactor would be to move the policy check away from the edge, closer to the database (where exactly, im not sure, as I'm not super familiar with the concourse codebase). Thoughts on how to proceed for this PR? I'd love to see this gap fixed in the next concourse release, but agree that the policy enforcement feature as a whole feels a bit half-baked. One larger change I'd like to see is a shift towards a deny based policy check, that allows for a list of violations to be returned and propagated downstream to the user. The current blanket policy failure error can be frustrating. Happy to take a stab at any of this if there's a preferred path forward.

Reply:
@vito @evanchaoli any thoughts on how to proceed?

Reply:
@cludden Sorry for the wait on this - just got back from vacation.

@andy-paine has split the above out into a separate discussion: https://github.com/concourse/concourse/discussions/6013 (thanks!)

For the sake of avoiding scope creep, I think we can review this PR as-is now that the larger issue has been split out into a separate discussion so that the context doesn't get lost. 👍 I'll try to wrap this up today!

It looks like this needs a rebase now though.

Reply:
I suspect the rebase is partly due to the "reasons" changes. Happy to help getting this sorted on top of that if needed.

Reply:
Updated PR with:
- rebase to pick up changes
- rename `RunSetPipeline` to `SetPipeline`
- refactored to support new policy check output (thanks for adding @andy-paine !)